Akka HTTP SSLEngine behaviour not supported in Java 16+

Problem Summary
==================
Warning on the server logs:

Illegal reflective access to method 
```
sun.security.ssl.SSLEngineImpl.setHandshakeApplicationProtocolSelector(java.util.function.BiFunction)
```
https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/SSLEngine.html#setHandshakeApplicationProtocolSelector-java.util.function.BiFunction-


this only happens on Java 11 (Adopt or GraalVM), but not on Java 8.

As mentioned on the warnings below, future versions of Java may remove this behaviour completely, so this could break in Java 17.

Logs:
--------
```bash
[info] TLSv1.3
[info] (gRPC server bound to {}:{},127.0.0.1,8080)
[error] WARNING: An illegal reflective access operation has occurred
[error] WARNING: Illegal reflective access by scala.reflect.package$ (file:/tmp/samples/lab-akka-grpc/target/bg-jobs/sbt_807ed737/target/7131136b/3588b5ba/scala-library-2.13.4.jar) to method sun.security.ssl.SSLEngineImpl.setHandshakeApplicationProtocolSelector(java.util.function.BiFunction)
[error] WARNING: Please consider reporting this to the maintainers of scala.reflect.package$
[error] WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
[error] WARNING: All illegal access operations will be denied in a future release
```

Steps to reproduce
-----------------------

Run both server and client on:
https://github.com/akka/akka-grpc-quickstart-scala.g8

Runtime
---------
```
scalaVersion := "2.13.4"
akkaVersion = "2.6.14"
akkaHttpVersion = "10.2.4"
akkaGrpcVersion = "1.1.1"

$ java -version
openjdk version "11.0.11" 2021-04-20
OpenJDK Runtime Environment AdoptOpenJDK-11.0.11+9 (build 11.0.11+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK-11.0.11+9 (build 11.0.11+9, mixed mode)

$ sw_vers
ProductName:	macOS
ProductVersion:	11.2.3
BuildVersion:	20D91
```

Notes
------

Server
Java 11 (AdoptJDK or GraalVM) fails
Java 8 (AdoptJDK) works

Client
Java 11 (AdoptJDK or GraalVM) works
Java 8 (AdoptJDK) works

I’ve tried combinations of these flags on the SBT build

```scala
fork := true
javaOptions := Seq("-Djdk.tls.server.protocols=TLSv1.3")
```

They didn’t make a difference. Neither did work trying to specify the TLS version on the server's SSLContext.

```
    val context = SSLContext.getInstance("TLSv1.3")
```

On Java 16
-----------
This is what happens when running on Java 16

```
 sbt runServer
copying runtime jar...
[info] welcome to sbt 1.5.1 (AdoptOpenJDK Java 16.0.1)
[info] running (fork) com.example.helloworld.GreeterServer 
[info] [2021-04-30 18:51:14,878] [INFO] [akka.event.slf4j.Slf4jLogger] [GreeterServer-akka.actor.default-dispatcher-3] [] - Slf4jLogger started
[info] TLSv1.3
[info] (gRPC server bound to {}:{},127.0.0.1,8443)
[info] [2021-04-30 18:51:33,740] [ERROR] [akka.actor.OneForOneStrategy] [GreeterServer-akka.actor.default-dispatcher-9] [akka://GreeterServer/system/Materializers/StreamSupervisor-0/TLS-for-flow-2-1] - Unable to make public void sun.security.ssl.SSLEngineImpl.setHandshakeApplicationProtocolSelector(java.util.function.BiFunction) accessible: module java.base does not "opens sun.security.ssl" to unnamed module @654b5005
[info] akka.actor.ActorInitializationException: akka://GreeterServer/system/Materializers/StreamSupervisor-0/TLS-for-flow-2-1: exception during creation
[info] 	at akka.actor.ActorInitializationException$.apply(Actor.scala:196)
[info] 	at akka.actor.ActorCell.create(ActorCell.scala:663)
[info] 	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:513)
[info] 	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:535)
[info] 	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:295)
[info] 	at akka.dispatch.Mailbox.run(Mailbox.scala:230)
[info] 	at akka.dispatch.Mailbox.exec(Mailbox.scala:243)
[info] 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:295)
[info] 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
[info] 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
[info] 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
[info] 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
[info] Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make public void sun.security.ssl.SSLEngineImpl.setHandshakeApplicationProtocolSelector(java.util.function.BiFunction) accessible: module java.base does not "opens sun.security.ssl" to unnamed module @654b5005
```

References
--------------

https://github.com/akka/akka-http/issues/3167
https://github.com/akka/akka/pull/29121
https://www.java.com/en/configure_crypto.html
https://docs.oracle.com/en/java/javase/16/migrate/removed-apis.html#GUID-FAD4E80D-64BA-42AC-A682-38D06EE61AC6
https://docs.oracle.com/en/java/javase/16/migrate/removed-tools-and-components.html#GUID-D7936F0D-08A9-411E-AD2F-E14A38DA56A7

--- a/akka-http-core/src/main/mima-filters/10.2.4.backwards.excludes/3811-http2-internals.excludes
+++ b/akka-http-core/src/main/mima-filters/10.2.4.backwards.excludes/3811-http2-internals.excludes
@@ -1,0 +1,1 @@
+ProblemFilters.exclude[Problem]("akka.http.impl.engine.http2.*")

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2AlpnSupport.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2AlpnSupport.scala
@@ -4,15 +4,14 @@
 
 package akka.http.impl.engine.http2
 
-import java.util.function.BiFunction
-import java.{ util => ju }
-
-import javax.net.ssl.SSLEngine
 import akka.annotation.InternalApi
+import akka.http.impl.engine.http2.Http2AlpnSupport.{ H2, HTTP11 }
 import akka.http.impl.util.JavaVersion
 import akka.stream.TLSProtocol.NegotiateNewSession
 import akka.stream.impl.io.TlsUtils
 
+import java.{ util => ju }
+import javax.net.ssl.SSLEngine
 import scala.util.Try
 
 /**
@@ -29,10 +28,14 @@ private[http] object Http2AlpnSupport {
    * Enables server-side Http/2 ALPN support for the given engine.
    */
   def enableForServer(engine: SSLEngine, setChosenProtocol: String => Unit): SSLEngine =
-    if (isAlpnSupportedByJDK) jdkAlpnSupport(engine, setChosenProtocol)
+    if (isAlpnSupportedByJDK) Http2JDKAlpnSupport.jdkAlpnSupport(engine, setChosenProtocol)
     else throw new RuntimeException(s"Need to run on a JVM >= 8u252 for ALPN support needed for HTTP/2. Running on ${sys.props("java.version")}")
 
-  def isAlpnSupportedByJDK: Boolean =
+  def clientSetApplicationProtocols(engine: SSLEngine, protocols: Array[String]): Unit =
+    if (isAlpnSupportedByJDK) Http2JDKAlpnSupport.clientSetApplicationProtocols(engine, protocols)
+    else throw new RuntimeException(s"Need to run on a JVM >= 8u252 for ALPN support needed for HTTP/2. Running on ${sys.props("java.version")}")
+
+  private def isAlpnSupportedByJDK: Boolean =
     // ALPN is supported starting with JDK 9
     JavaVersion.majorVersion >= 9 ||
       (classOf[SSLEngine].getMethods.exists(_.getName == "setHandshakeApplicationProtocolSelector")
@@ -44,12 +47,17 @@ private[http] object Http2AlpnSupport {
           if (jettyAlpnClassesAvailable) throw new RuntimeException("On JDK >= 8u252 you need to either remove jetty-alpn-agent or use version 2.0.10 (which is a noop)")
           else true
         })
+}
 
-  private type SSLEngineWithALPNSupport = {
-    def setHandshakeApplicationProtocolSelector(selector: BiFunction[SSLEngine, ju.List[String], String]): Unit
-  }
+/**
+ * INTERNAL API
+ *
+ * The actual implementation of ALPN support on supported JDKs. We rely on lazy class loading to not fail with class loading errors
+ * when ALPN support is missing.
+ */
+private[http] object Http2JDKAlpnSupport {
   def jdkAlpnSupport(engine: SSLEngine, setChosenProtocol: String => Unit): SSLEngine = {
-    engine.asInstanceOf[SSLEngineWithALPNSupport].setHandshakeApplicationProtocolSelector { (engine: SSLEngine, protocols: ju.List[String]) =>
+    engine.setHandshakeApplicationProtocolSelector { (engine: SSLEngine, protocols: ju.List[String]) =>
       val chosen = chooseProtocol(protocols)
       chosen.foreach(setChosenProtocol)
 
@@ -68,13 +76,9 @@ private[http] object Http2AlpnSupport {
 
   def applySessionParameters(engine: SSLEngine, sessionParameters: NegotiateNewSession): Unit = TlsUtils.applySessionParameters(engine, sessionParameters)
 
-  private type SSLParametersWithALPNSupport = {
-    def setApplicationProtocols(protocols: Array[String]): Unit
-  }
   def clientSetApplicationProtocols(engine: SSLEngine, protocols: Array[String]): Unit = {
     val params = engine.getSSLParameters
-    params.asInstanceOf[SSLParametersWithALPNSupport].setApplicationProtocols(Array("h2"))
+    params.setApplicationProtocols(Array("h2"))
     engine.setSSLParameters(params)
   }
-
 }

